File: Analysis/smart_graph.py
import json
from pathlib import Path
from typing import List, Dict, Any
from Core.types import FunctionRecord, SmellIssue, DuplicateGroup, Severity


class SmartGraph:
    """Generates a visualization of the codebase health and relationships (Premium Edition)."""

    def __init__(self):
        self.nodes: List[Dict[str, Any]] = []
        self.edges: List[Dict[str, Any]] = []

    def build(
        self,
        functions: List[FunctionRecord],
        smells: List[SmellIssue],
        duplicates: List[DuplicateGroup],
        root: Path,
        mode: str = "complete"  # 'complete', 'calls', 'hierarchy'
    ):
        """Build the graph nodes and edges."""
        self.nodes = []
        self.edges = []
        smell_map = self._build_smell_map(smells)

        if mode == "calls":
            self.nodes, self.edges = self._build_call_graph(functions, smell_map)
        elif mode == "hierarchy":
            self.nodes, self.edges = self._build_hierarchy_graph(functions, root)
        else:  # default complete graph
            self.nodes = [self._make_node(f, smell_map.get(f.key, [])) for f in functions]
            self.edges = self._make_edges(duplicates)

    # -- private helpers (extracted from build) ------------------------------

    @staticmethod
    def _build_smell_map(smells: List[SmellIssue]) -> Dict[str, List[SmellIssue]]:
        """Map ``file::name`` keys to their associated SmellIssue lists."""
        smell_map: Dict[str, List[SmellIssue]] = {}
        for s in smells:
            if not s.name:
                continue
            p = Path(s.file_path)
            parent = str(p.parent).replace("\\", "/")
            stem = p.stem
            k = f"{stem}::{s.name}" if parent == "." else f"{parent}/{stem}::{s.name}"
            smell_map.setdefault(k, []).append(s)
        return smell_map

    @staticmethod
    def _make_node(f: FunctionRecord, f_smells: List[SmellIssue]) -> Dict[str, Any]:
        """Create a single graph node dict for *f*."""
        critical_count = sum(1 for s in f_smells if s.severity == Severity.CRITICAL)
        warning_count = sum(1 for s in f_smells if s.severity == Severity.WARNING)

        if critical_count > 0:
            color, health = "#ff4d4f", "critical"
        elif warning_count > 0:
            color, health = "#faad14", "warning"
        else:
            color, health = "#52c41a", "healthy"

        issues_data = [{"category": s.category, "severity": s.severity, "message": s.message} for s in f_smells]
        return {
            "id": f.key,
            "label": f.name,
            "title": f.name,
            "color": color,
            "health": health,
            "size": max(10, f.size_lines),
            "group": str(Path(f.file_path).parent),
            "signature": f.name + "(" + ", ".join(f.parameters) + ")",
            "file": f.file_path,
            "line": f.line_start,
            "code": f.code[:1000] + ("..." if len(f.code) > 1000 else ""),
            "issues": issues_data,
            "complexity": f.complexity,
            "shape": "dot"
        }

    @staticmethod
    def _make_edges(duplicates: List[DuplicateGroup]) -> List[Dict[str, Any]]:
        """Generate pairwise edge dicts from DuplicateGroups."""
        edges: List[Dict[str, Any]] = []
        for group in duplicates:
            funcs = group.functions
            if len(funcs) < 2:
                continue
            for i in range(len(funcs)):
                for j in range(i + 1, len(funcs)):
                    edges.append(
                        {
                            "from": funcs[i]["key"],
                            "to": funcs[j]["key"],
                            "value": group.avg_similarity,
                            "title": f"{group.similarity_type} duplicate ({group.avg_similarity:.2f})",
                            "color": {"color": "#ffffff", "opacity": 0.3},
                            "dashes": True,
                        }
                    )
        return edges

    def _build_call_graph(self, functions: List[FunctionRecord], smell_map: Dict):
        nodes = []
        edges = []
        func_keys = {f.name: f.key for f in functions}
        for f in functions:
            nodes.append(self._make_node(f, smell_map.get(f.key, [])))
            for call in f.calls_to:
                if call in func_keys:
                    edges.append({
                        "from": f.key,
                        "to": func_keys[call],
                        "arrows": "to",
                        "color": {"color": "#a855f7", "opacity": 0.6},
                        "title": f"calls {call}"
                    })
        return nodes, edges

    def _build_hierarchy_graph(self, functions: List[FunctionRecord], root: Path):
        nodes = []
        edges = []
        dirs = set()
        files = set()

        nodes.append({"id": "root", "label": root.name, "color": "#1e293b", "size": 30, "shape": "box", "font": {"color": "white"}})

        for f in functions:
            nodes.append(self._make_node(f, []))
            p = Path(f.file_path)
            parts = list(p.parts)

            parent_id = "root"
            current_path = ""
            for i, part in enumerate(parts):
                current_path = current_path + "/" + part if current_path else part
                is_file = (i == len(parts) - 1)
                node_id = f"path:{current_path}"

                if current_path not in dirs and current_path not in files:
                    if is_file:
                        files.add(current_path)
                        nodes.append({"id": node_id, "label": part, "color": "#334155", "size": 20, "shape": "box", "font": {"color": "white"}})
                    else:
                        dirs.add(current_path)
                        nodes.append({"id": node_id, "label": part, "color": "#0f172a", "size": 25, "shape": "box", "font": {"color": "white"}})

                    edges.append({"from": parent_id, "to": node_id, "color": {"color": "#64748b", "opacity": 0.4}, "arrows": "to"})
                parent_id = node_id

            edges.append({"from": parent_id, "to": f.key, "color": {"color": "#94a3b8", "opacity": 0.4}})

        return nodes, edges

    def write_html(self, output_path: Path):
        """Not used by GUI, but keeping signature for CLI compatibility."""
        output_path.write_text("<html><body>Use GUI to view premium graph.</body></html>", encoding="utf-8")


# Module-level API for test compatibility
_default_analyzer = SmartGraph()

def build(*args, **kwargs):
    """Wrapper for SmartGraph.build()."""
    return _default_analyzer.build(*args, **kwargs)

def write_html(*args, **kwargs):
    """Wrapper for SmartGraph.write_html()."""
    return _default_analyzer.write_html(*args, **kwargs)

File: X_Ray_Rust_Full/src/smart_graph.rs
// src/smart_graph.rs — Interactive HTML codebase health visualization
//
// Port of Analysis/smart_graph.py
// Builds a node/edge graph of functions colored by health with duplicate links,
// then exports an interactive HTML page using vis-network.js.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;

use crate::types::{DuplicateGroup, FunctionRecord, Severity, SmellIssue};

/// A graph node representing a single function.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphNode {
    pub id: String,
    pub label: String,
    pub title: String,
    pub color: String,
    pub health: String,
    pub size: u32,
    pub group: String,
}

/// A graph edge linking duplicate/similar functions.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphEdge {
    pub from: String,
    pub to: String,
    pub value: f64,
    pub title: String,
}

/// Build the full graph from scan results.
pub fn build_graph(
    functions: &[FunctionRecord],
    smells: &[SmellIssue],
    duplicates: &[DuplicateGroup],
) -> (Vec<GraphNode>, Vec<GraphEdge>) {
    let smell_map = build_smell_map(smells);

    let nodes: Vec<GraphNode> = functions
        .iter()
        .map(|f| {
            let key = f.key();
            let f_smells = smell_map.get(&key).cloned().unwrap_or_default();
            make_node(f, &f_smells)
        })
        .collect();

    let edges = make_edges(duplicates);

    (nodes, edges)
}

/// Map `file_stem::name` keys to their associated SmellIssue indices.
fn build_smell_map(smells: &[SmellIssue]) -> HashMap<String, Vec<SmellIssue>> {
    let mut map: HashMap<String, Vec<SmellIssue>> = HashMap::new();
    for s in smells {
        if s.name.is_empty() {
            continue;
        }
        let p = Path::new(&s.file_path);
        let parent = p
            .parent()
            .map(|pp| pp.to_string_lossy().replace('\\', "/"))
            .unwrap_or_default();
        let stem = p
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default();

        let key = if parent == "." || parent.is_empty() {
            format!("{}::{}", stem, s.name)
        } else {
            format!("{}/{}::{}", parent, stem, s.name)
        };
        map.entry(key).or_default().push(s.clone());
    }
    map
}

/// Create a single graph node for a function.
fn make_node(f: &FunctionRecord, f_smells: &[SmellIssue]) -> GraphNode {
    let critical_count = f_smells
        .iter()
        .filter(|s| matches!(s.severity, Severity::Critical))
        .count();
    let warning_count = f_smells
        .iter()
        .filter(|s| matches!(s.severity, Severity::Warning))
        .count();

    let (color, health) = if critical_count > 0 {
        ("#e74c3c", "critical")
    } else if warning_count > 0 {
        ("#f39c12", "warning")
    } else {
        ("#2ecc71", "healthy")
    };

    let mut tooltip = format!(
        "<b>{}</b><br>{}:{}<br>",
        f.name, f.file_path, f.line_start
    );
    if !f_smells.is_empty() {
        tooltip.push_str("<br><b>Issues:</b><br>");
        for s in f_smells {
            let icon = s.severity.icon();
            tooltip.push_str(&format!("{} {}: {}<br>", icon, s.category, s.message));
        }
    }

    let group = Path::new(&f.file_path)
        .parent()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_default();

    GraphNode {
        id: f.key(),
        label: f.name.clone(),
        title: tooltip,
        color: color.to_string(),
        health: health.to_string(),
        size: f.size_lines,
        group,
    }
}

/// Generate pairwise edges from DuplicateGroups.
fn make_edges(duplicates: &[DuplicateGroup]) -> Vec<GraphEdge> {
    let mut edges = Vec::new();
    for group in duplicates {
        if group.functions.len() < 2 {
            continue;
        }
        for i in 0..group.functions.len() {
            for j in (i + 1)..group.functions.len() {
                let key_from = group.functions[i]
                    .get("key")
                    .and_then(|k| k.as_str())
                    .unwrap_or("")
                    .to_string();
                let key_to = group.functions[j]
                    .get("key")
                    .and_then(|k| k.as_str())
                    .unwrap_or("")
                    .to_string();

                if !key_from.is_empty() && !key_to.is_empty() {
                    edges.push(GraphEdge {
                        from: key_from,
                        to: key_to,
                        value: group.avg_similarity,
                        title: format!(
                            "{} duplicate ({:.2})",
                            group.similarity_type, group.avg_similarity
                        ),
                    });
                }
            }
        }
    }
    edges
}

/// Export the graph to an interactive HTML file using vis-network.
pub fn write_html(
    nodes: &[GraphNode],
    edges: &[GraphEdge],
    output_path: &Path,
) -> std::io::Result<()> {
    let nodes_json = serde_json::to_string(nodes).unwrap_or_else(|_| "[]".to_string());
    let edges_json = serde_json::to_string(edges).unwrap_or_else(|_| "[]".to_string());

    let html = format!(
        r##"<!DOCTYPE html>
<html>
<head>
    <title>X-RAY Codebase Visualization</title>
    <script type="text/javascript" src="https://unpkg.com/vis-network/standalone/umd/vis-network.min.js"></script>
    <style type="text/css">
        body {{ font-family: 'Segoe UI', Tahoma, sans-serif; margin: 0; padding: 20px; background: #1a1a2e; color: #eee; }}
        h2 {{ color: #00d4ff; }}
        #legend {{ margin: 10px 0; }}
        #legend span {{ margin-right: 18px; }}
        .dot {{ display: inline-block; width: 12px; height: 12px; border-radius: 50%; margin-right: 4px; vertical-align: middle; }}
        #mynetwork {{ width: 100%; height: 85vh; border: 1px solid #333; border-radius: 8px; background: #16213e; }}
        #stats {{ margin-top: 8px; color: #aaa; font-size: 0.9em; }}
    </style>
</head>
<body>
    <h2>X-RAY — Codebase Health Graph</h2>
    <div id="legend">
        <span><span class="dot" style="background:#2ecc71"></span>Healthy</span>
        <span><span class="dot" style="background:#f39c12"></span>Warning</span>
        <span><span class="dot" style="background:#e74c3c"></span>Critical</span>
    </div>
    <div id="mynetwork"></div>
    <div id="stats"></div>
    <script type="text/javascript">
        var nodesData = {nodes_json};
        var edgesData = {edges_json};
        var nodes = new vis.DataSet(nodesData);
        var edges = new vis.DataSet(edgesData);
        var container = document.getElementById('mynetwork');
        var data = {{ nodes: nodes, edges: edges }};
        var options = {{
            nodes: {{
                shape: 'dot',
                font: {{ size: 14, color: '#ddd' }},
                scaling: {{ min: 8, max: 40, label: {{ enabled: true, min: 10, max: 20 }} }}
            }},
            edges: {{
                color: {{ color: '#555', highlight: '#00d4ff' }},
                smooth: false
            }},
            physics: {{
                stabilization: false,
                barnesHut: {{
                    gravitationalConstant: -80000,
                    springConstant: 0.001,
                    springLength: 200
                }}
            }},
            interaction: {{ hover: true, tooltipDelay: 100 }}
        }};
        var network = new vis.Network(container, data, options);
        // Stats
        var healthy = nodesData.filter(n => n.health === 'healthy').length;
        var warning = nodesData.filter(n => n.health === 'warning').length;
        var critical = nodesData.filter(n => n.health === 'critical').length;
        document.getElementById('stats').innerHTML =
            'Functions: ' + nodesData.length + ' | Edges: ' + edgesData.length +
            ' | Healthy: ' + healthy + ' | Warning: ' + warning + ' | Critical: ' + critical;
    </script>
</body>
</html>"##
    );

    // Ensure parent dir exists
    if let Some(parent) = output_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(output_path, html)?;
    Ok(())
}
